Fix issue #4476 related to BatchedDot.

notoraptor opened the pull request:
fix gh-4476. It is fixed by modifying contiguousity checking in the c_code of BatchedDot. Now the code uses `(PyArray_IS_C_CONTIGUOUS || PyArray_IS_F_CONTIGUOUS)` to do the checking.

Previously the code checked if every stride of the inputs is relevant (strictly positive and multiple of type size) and if at least one of the strides has the size of the type. I think the bug comes from the latter condition, as an array can be not-contiguous even if one stride has the size of the type (for example, a Fortran array in which the first dimension is contiguous but not the last ones).

Remarks:
* I put the test in `test_basic` even if `BatchedDot` is in `blas.py` because the `batched_dot` tests are also in `test_basic`.
* It seems some other codes check the contiguity using the previous conditions, so there is maybe some similar bugs with other ops. For example, we may have to check the Dot22 op: https://github.com/Theano/Theano/blob/master/theano/tensor/blas.py#L603

@lamblin @nouiz !

nouiz commented:
Squash the last commit into the first one to prevent a c code cache version problem in the history.

nouiz commented on theano/tensor/tests/test_basic.py:
```diff
@@ -2771,6 +2771,30 @@ def test_batched_dot():
+
+    reversed_x_container = np_genarray(20, 40, 30)
+    x_container = reversed_x_container.T
+    x = x_container[::1, ::2, ::2]
```
Add a test with x_container[::-1, ::2, ::2]

nouiz commented on theano/tensor/blas.py:
```diff
@@ -2149,14 +2149,8 @@ def c_code(self, node, name, inp, out, sub):
-                                     .format(strides=strides, i=i) for i in range(ndim)),
-            ])
+        def contiguous(var):
+            return "(PyArray_IS_C_CONTIGUOUS(%s) || PyArray_IS_F_CONTIGUOUS(%s))" % (var, var)
```
This is too string. I think you could just change the range(ndim) to range(1, ndim) in the previous code.

notoraptor commented:
Updated. I put back the original checking and I updated it to make it more precise.

New tests are also updated.

Tests passed:

```
nosetests -vs theano/tensor/tests/test_basic.py:test_batched_dot
nosetests -vs theano/tensor/tests/test_basic.py:test_batched_dot_not_contiguous
nosetests -vs theano/tensor/tests/test_basic.py:BatchedDotTester
```

nouiz commented on theano/tensor/blas.py:
```diff
@@ -2154,8 +2154,8 @@ def contiguous(var, ndim):
-                "(%s)" % " || ".join("{strides}[{i}] == type_size"
-                                     .format(strides=strides, i=i) for i in range(ndim)),
+                "(%s)" % (" || ".join("{strides}[{i}] == type_size"
+                                      .format(strides=strides, i=i) for i in range(1, ndim)) or '1'),
```
why the `or '1'`? Also I think both range should be `range(1, ndim)`

nouiz commented:
jenkins failure are not related to this PR and I have another PR that fix them. So merging.
